### PR TITLE
Feat/spring security

### PR DESCRIPTION
**Pull request by NickKlock**

### CODE REVIEW

File: backend/src/main/java/de/neuefische/backend/controller/MongoUserController.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+import de.neuefische.backend.model.Traveller;
+import de.neuefische.backend.service.MongoUserService;
+import org.springframework.security.core.context.SecurityContextHolder;
+import org.springframework.web.bind.annotation.*;
+
+import javax.servlet.http.HttpSession;
+import java.security.Principal;
+
+@RestController
+@RequestMapping("/api/user")
```
Comment (s-gbz) [inline backend/src/main/java/de/neuefische/backend/controller/MongoUserController.java:13]: Good habit to use `api/`!

File: backend/src/main/java/de/neuefische/backend/controller/MongoUserController.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    public String helloMe(Principal principal) {
+        if (principal != null) {
+            return principal.getName();
+        }
+        return "anonymousUser";
+    }
+
+    @PostMapping("/login")
+    public String login() {
+        //lets give back a nice object.
```
Comment (s-gbz) [inline backend/src/main/java/de/neuefische/backend/controller/MongoUserController.java:41]: Why don't you?

File: backend/src/main/java/de/neuefische/backend/controller/MongoUserController.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+    @PostMapping("/logout")
+    public String logout(HttpSession httpSession) {
+        httpSession.invalidate();
+        SecurityContextHolder.clearContext();
+        return "anonymousUser";
+    }
+
+
+    @GetMapping("/{name}")
```
Comment (s-gbz) [inline backend/src/main/java/de/neuefische/backend/controller/MongoUserController.java:56]: Couldn't that escalate into a huge security issue?

File: backend/src/main/java/de/neuefische/backend/controller/TravellerController.java [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+import de.neuefische.backend.model.Traveller;
+import de.neuefische.backend.model.TravellerRequest;
+import de.neuefische.backend.service.TravellerService;
+import org.springframework.web.bind.annotation.*;
+
+import java.util.List;
+
+@RestController
+@RequestMapping("/api/traveller")
```
Comment (s-gbz) [inline backend/src/main/java/de/neuefische/backend/controller/TravellerController.java:11]:
```suggestion
@RequestMapping("/api/travellers")
```
I'd suggest using plural nouns to be more RESTful.
It's definetely debatable, but I tend to see [some favor](https://stackoverflow.com/questions/6845772/should-i-use-singular-or-plural-name-convention-for-rest-resources) towards the plural convention and find it more logical.


(And yes, I noticed that @danielschwarz-neuefische seems to be in "team-single-nouns" ;D - Let's chat about it next week? :coffee: )

File: backend/src/main/java/de/neuefische/backend/controller/MongoUserController.java (hunk 1 above)
Comment (s-gbz) [reply backend/src/main/java/de/neuefische/backend/controller/MongoUserController.java:13]: Oh, didn't catch it first, see my comment below towards RESTfulness.

File: backend/src/main/java/de/neuefische/backend/controller/TravellerController.java [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+
+    @GetMapping()
+    public List<Traveller> listTravellers(){
+        return travellerService.listTravellers();
+    }
+
+
+    @PutMapping("/{id}")
+    public Traveller update(@PathVariable String id, @RequestBody TravellerRequest travellerRequest){
+        Traveller toEditUser = new Traveller(id,travellerRequest.name(),travellerRequest.visitedCountries());
```
Comment (s-gbz) [inline backend/src/main/java/de/neuefische/backend/controller/TravellerController.java:28]: Isn't it more of a `travellerToUpdate` (or `toEdit` if you like)?

File: backend/src/main/java/de/neuefische/backend/exception/NoSuchTravellerException.java [hunk 6]
```diff
@@ -4,8 +4,8 @@
 import org.springframework.web.bind.annotation.ResponseStatus;
 
 @ResponseStatus(value = HttpStatus.NOT_FOUND)
-public class NoSuchUserException extends IllegalArgumentException {
-    public NoSuchUserException(String s) {
+public class NoSuchTravellerException extends IllegalArgumentException {
```
Comment (s-gbz) [inline backend/src/main/java/de/neuefische/backend/exception/NoSuchTravellerException.java:7]: :exploding_head: 

Custom exception

:exploding_head: 

Custom ResponseStatus

File: backend/src/main/java/de/neuefische/backend/model/MongoUserRequest.java [hunk 7]
```diff
@@ -0,0 +1,7 @@
+package de.neuefische.backend.model;
+
+public record MongoUserRequest(
```
Comment (s-gbz) [inline backend/src/main/java/de/neuefische/backend/model/MongoUserRequest.java:3]: Nice

File: backend/src/main/java/de/neuefische/backend/security/MongoUserDetailsService.java [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+    private final MongoUserRepo repo;
+
+    public MongoUserDetailsService(MongoUserRepo repo) {
+        this.repo = repo;
+    }
+
+    @Override
+    public UserDetails loadUserByUsername(String username) throws UsernameNotFoundException {
+        MongoUser user = repo.findByUsername(username)
+                .orElseThrow(()-> new UsernameNotFoundException("Bad Info"));
```
Comment (s-gbz) [inline backend/src/main/java/de/neuefische/backend/security/MongoUserDetailsService.java:23]: Why not include the username that was requested?

File: backend/src/main/java/de/neuefische/backend/security/SecurityConfig.java [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+
+    @Bean
+    public PasswordEncoder passwordEncoder() {
+        return new Argon2PasswordEncoder();
+    }
+
+    @Bean
+    public SecurityFilterChain filterChain(HttpSecurity http) throws Exception {
+        return http
+                .csrf().disable()
```
Comment (s-gbz) [inline backend/src/main/java/de/neuefische/backend/security/SecurityConfig.java:25]: It's fine, you can ignore Sonarlint here. Will fix it tomorrow.
However, why not include the fix then (tomorrow) :)

File: backend/src/main/java/de/neuefische/backend/service/MongoUserService.java [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+import org.springframework.security.core.userdetails.UsernameNotFoundException;
+import org.springframework.security.crypto.argon2.Argon2PasswordEncoder;
+import org.springframework.security.crypto.password.PasswordEncoder;
+import org.springframework.stereotype.Service;
+
+import java.util.HashSet;
+import java.util.Optional;
+
+@Service
+public class MongoUserService {
```
Comment (s-gbz) [inline backend/src/main/java/de/neuefische/backend/service/MongoUserService.java:17]: Great job combining `Traveller` & `MongoUser`

File: backend/src/main/java/de/neuefische/backend/service/TravellerService.java [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+    }
+
+    public List<Traveller> listTravellers() {
+        return travellerRepo.findAll();
+    }
+
+
+    public Traveller findTravellerById(String id){
+        return travellerRepo.findById(id)
+                .orElseThrow(() -> new NoSuchElementException("User with id " + id + "does not exist"));
```
Comment (s-gbz) [inline backend/src/main/java/de/neuefische/backend/service/TravellerService.java:26]: Didn't you create a custom exception already? Use it :wink:

File: backend/src/main/java/de/neuefische/backend/service/TravellerService.java [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+    }
+
+    public Traveller updateTraveller(String id, Traveller editedUser){
+        checkIfTravellerExists(id);
+        return travellerRepo.save(editedUser);
+    }
+
+    private void checkIfTravellerExists(String id) throws NoSuchTravellerException {
```
Comment (s-gbz) [inline backend/src/main/java/de/neuefische/backend/service/TravellerService.java:34]: I like this A LOT.
Clean. Code.

File: backend/src/test/java/de/neuefische/backend/controller/MongoUserControllerTest.java [hunk 13]
```diff
@@ -0,0 +1,106 @@
...
+import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.*;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.content;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
+
+@SpringBootTest
+@AutoConfigureMockMvc
+@TestInstance(TestInstance.Lifecycle.PER_CLASS)
+@DirtiesContext(classMode = DirtiesContext.ClassMode.AFTER_EACH_TEST_METHOD)
+class MongoUserControllerTest {
+    String userEndPoint = "/api/user";
```
Comment (s-gbz) [inline backend/src/test/java/de/neuefische/backend/controller/MongoUserControllerTest.java:30]: :sweat_drops:

File: backend/src/test/java/de/neuefische/backend/controller/MongoUserControllerTest.java [hunk 14]
```diff
@@ -0,0 +1,106 @@
...
+    @BeforeAll
+    public void setup(){
+        mvc = MockMvcBuilders
+                .webAppContextSetup(context)
+                .apply(SecurityMockMvcConfigurers.springSecurity()).
+                build();
+    }
+
+    @Test
+    void addUser_expect_traveller_status_isOk() throws Exception {
```
Comment (s-gbz) [inline backend/src/test/java/de/neuefische/backend/controller/MongoUserControllerTest.java:48]: You guys <3

File: backend/src/test/java/de/neuefische/backend/controller/MongoUserControllerTest.java [hunk 15]
```diff
@@ -0,0 +1,106 @@
...
+                .getContentAsString();
+
+        Traveller resultTraveller = objectMapper.readValue(travellerResult, Traveller.class);
+        assertEquals("nick",resultTraveller.name());
+
+    }
+
+    @Test
+    void delete_expect_status_isOk() throws Exception {
+        mvc.perform(delete(userEndPoint+"/1")
```
Comment (s-gbz) [inline backend/src/test/java/de/neuefische/backend/controller/MongoUserControllerTest.java:70]: How would you have successful deletion if there's no data?

Comment (s-gbz) [reply backend/src/test/java/de/neuefische/backend/controller/MongoUserControllerTest.java:70]: Each test case should be isolated

File: backend/src/test/java/de/neuefische/backend/controller/TravellerControllerTest.java [hunk 16]
```diff
@@ -0,0 +1,69 @@
...
+    public void setup(){
+        mvc = MockMvcBuilders
+                .webAppContextSetup(context)
+                .apply(SecurityMockMvcConfigurers.springSecurity()).
+                build();
+    }
+
+    @WithMockUser("spring")
+    @Test
+    void listUsers_expect_empty_list() throws Exception {
```
Comment (s-gbz) [inline backend/src/test/java/de/neuefische/backend/controller/TravellerControllerTest.java:46]: I'd like to see a test with actual data.
Tip: Use the actual repository if you'd like to add data hassle-free.

File: backend/src/test/java/de/neuefische/backend/security/MongoUserDetailsServiceTest.java [hunk 17]
```diff
@@ -0,0 +1,37 @@
...
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+class MongoUserDetailsServiceTest {
+    MongoUserRepo mongoUserRepo = mock(MongoUserRepo.class);
+
+    MongoUserDetailsService mongoUserDetailsService = new MongoUserDetailsService(mongoUserRepo);
+    @Test
+    void loadUserByUsername_expect_exception() {
+        assertThrows(UsernameNotFoundException.class, () -> mongoUserDetailsService.loadUserByUsername("test"));
```
Comment (s-gbz) [inline backend/src/test/java/de/neuefische/backend/security/MongoUserDetailsServiceTest.java:22]: :cool:

File: backend/src/test/java/de/neuefische/backend/service/MongoUserServiceTest.java [hunk 18]
```diff
@@ -0,0 +1,73 @@
...
+import static org.mockito.Mockito.when;
+
+class MongoUserServiceTest {
+    MongoUserRepo mongoUserRepo = mock(MongoUserRepo.class);
+    TravellerRepo travellerRepo = mock(TravellerRepo.class);
+    UuidGeneratorService uuidGeneratorService = mock(UuidGeneratorService.class);
+    MongoUserService mongoUserService = new MongoUserService(travellerRepo, uuidGeneratorService, mongoUserRepo);
+
+    @Test
+    void addNewUserAndTraveller_expect_givenId() {
```
Comment (s-gbz) [inline backend/src/test/java/de/neuefische/backend/service/MongoUserServiceTest.java:25]: What I really like about this, is how you structured the blocks into _blocks of thought_ :thought_balloon:

File: backend/src/test/java/de/neuefische/backend/service/MongoUserServiceTest.java [hunk 19]
```diff
@@ -0,0 +1,73 @@
...
+        MongoUserRequest givenMongoUserRequest = new MongoUserRequest("nick","pw");
+        String givenId = uuidGeneratorService.generateUuid();
+        Traveller givenTraveller = mongoUserService.createTraveller(givenId, givenMongoUserRequest);
+
+        when(mongoUserRepo.save(mongoUserService.createMongoUser(givenId, givenMongoUserRequest)))
+                .thenReturn(mongoUserService.createMongoUser(givenId, givenMongoUserRequest));
```
Comment (s-gbz) [inline backend/src/test/java/de/neuefische/backend/service/MongoUserServiceTest.java:31]: what

File: backend/src/test/java/de/neuefische/backend/service/MongoUserServiceTest.java [hunk 20]
```diff
@@ -0,0 +1,73 @@
...
+        String result = mongoUserService.deleteUserAndTravellerById(givenId);
+
+        assertEquals(givenId,result);
+
+    }
+
+    @Test
+    void getIdByUsername_expect_same_id() {
+        String givenName = "nick";
+        MongoUser toReturnUser = new MongoUser("0","nick","123");
```
Comment (s-gbz) [inline backend/src/test/java/de/neuefische/backend/service/MongoUserServiceTest.java:54]: I'm starting to see a pattern.
Tip: Don't start names with `to`.

File: frontend/src/App.tsx [hunk 21]
```diff
@@ -1,37 +1,38 @@
...
+import Login from "./component/Login";
+import useLoggedInUser from "./hook/useLoggedInUser";
+import ProtectedRoutes from "./component/ProtectedRoutes";
 
 
 function App() {
-    const [loggedInUser, setLoggedInUser] = useState<User>({id: "null", name: "loading", visitedCountries: []})
+
     const {countries} = useCountries()
+    const {loggedInTraveller, loginUser, username} = useLoggedInUser()
```
Comment (s-gbz) [inline frontend/src/App.tsx:17]: Why would you receive a traveller if everything else in the line references a user?

File: frontend/src/App.tsx [hunk 22]
```diff
@@ -1,37 +1,38 @@
...
-                    </Route>
-                    <Route path="/overview/:id" element={<p>detail</p>}></Route>
-                    <Route path="/overview/:id/profile" element={<p>edit name</p>}></Route>
-                    <Route path="/overview/:id/countries" element={<AddVisitCountry
-                        countries={countries}
-                        loggedInUser={loggedInUser}
-                    />}>
+                    <Route path="/signup" element={<SignUp/>}/>
+                    <Route path="/login" element={<Login handleLogInUser={loginUser} />}/>
+                    <Route element={<ProtectedRoutes loggedInTraveller={loggedInTraveller} username={username}/>}>
```
Comment (s-gbz) [inline frontend/src/App.tsx:27]: A simple blank line does wonders for readability.

File: frontend/src/apiCalls.ts [hunk 23]
```diff
@@ -1,26 +1,52 @@
...
+export function login(toLoginUser: LoginUser) {
+    return axios.post("/api/user/login", undefined, {
+        auth: {
+            username:toLoginUser.username,
+            password:toLoginUser.password
+        }
+    }).then(response => response.data)
+}
+
+export function getLoggedUserName(){
```
Comment (s-gbz) [inline frontend/src/apiCalls.ts:37]:
```suggestion
export function getLoggedInUserName(){
```

File: frontend/src/component/AddUser.tsx [hunk 24]
```diff
@@ -29,12 +30,24 @@ export default function AddUser(props:AddUserProps){
...
                         placeholder={"hi, traveller"}
                     />
                 </label>
+
+                <label>
+                    Password:
+                    <input
+                        type={"password"}
+                        name={"password"}
+                        value={newUser.password}
```
Comment (s-gbz) [inline frontend/src/component/AddUser.tsx:45]: A shared state for username and pw - "[you make me happy :guitar:](https://www.youtube.com/watch?v=fFtZU2Aka80)"

File: frontend/src/component/UserOverview.tsx [hunk 25]
```diff
@@ -33,8 +31,8 @@ export default function UserOverview(props: UserOverviewProps) {
         setSearchQuery(event.target.value)
     }
 
-    function handleEditUserName(id: string, user: User) {
-        props.setLoggedInUser(user)
+    function handleEditUserName(id: string, user: Traveller) {
```
Comment (s-gbz) [inline frontend/src/component/UserOverview.tsx:34]: User... Traveller... What is it now?

File: frontend/src/model/User.ts [hunk 26]
```diff
@@ -1,11 +1,17 @@
...
     visitedCountries: Country[]
 }
 
 export type NewUser = {
-    name : string,
+    password: string;
+    username: string,
+}
+
+export type LoginUser = {
```
Comment (s-gbz) [inline frontend/src/model/User.ts:14]: Perhaps `UserLoginRequest`?

**Review by s-gbz — COMMENTED** [review]
So, commit messages are getting more expressive - good job.
Some FYI for additional [knowledge](https://www.youtube.com/watch?v=jksPhQhJRoc) :tm: : You can use GitKraken to easily squash countless "Fix test" commits or also use `git commit --amend ` and then force-push those fixes (or simply wait a bit before pushing changes to the server)

Jesus. Christ. You killed me :gun: 
Super nice work, HOWWWWWWWWWWWWWWWWWEVER.
(At least) 2. PR's. ok?
Because "[hier geht's nicht mehr um so was](https://www.youtube.com/watch?v=x-33OUv0HhA)", this is refactoring, expansion AND security. So, two, at least one slices to many :wink: 

Make sure you find all my easter eggs and let's have a chat tomorrow if there's any need for discussion. And again, amazing work team 4 :heart: (you're team 4 right?) 

PS: Did actually ALL of you participate in coding though :monocle_face: :wink:  ?

File: backend/src/main/java/de/neuefische/backend/security/MongoUserDetailsService.java (hunk 8 above)
Comment (NickKlock) [reply backend/src/main/java/de/neuefische/backend/security/MongoUserDetailsService.java:23]: It's bad practice and also a security issue. U never want to return the actual username in error responses

**Review by s-gbz — APPROVED** [review]

**Review by LilianCharlotte — APPROVED** [review]

**Pull request merged**